File: app/src/main/java/com/qw/recyclerview/sample/SwipeRefreshLayoutPlusActivity.java
package com.qw.recyclerview.sample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.qw.recyclerview.core.BaseViewHolder;
import com.qw.recyclerview.core.State;
import com.qw.recyclerview.sample.core.BaseSwipeRefreshLayoutListActivity;
import com.qw.recyclerview.sample.databinding.SwipeRefreshLayoutActivityBinding;

public class SwipeRefreshLayoutPlusActivity extends BaseSwipeRefreshLayoutListActivity<String> {
    private SwipeRefreshLayoutActivityBinding bind;

    @Override
    protected void setContentView() {
        bind = SwipeRefreshLayoutActivityBinding.inflate(getLayoutInflater());
        setContentView(bind.getRoot());
    }

    @Override
    protected void initData(Bundle savedInstanceState) {
       smartRefreshHelper.autoRefresh();
    }

    @Override
    public void onRefresh() {
        super.onRefresh();
        smartRefreshHelper.getRecyclerView().postDelayed(new Runnable() {
            @Override
            public void run() {
                modules.clear();
                for (int i = 0; i < 20; i++) {
                    modules.add("" + i);
                }
                adapter.notifyDataSetChanged();
                adapter.notifyFooterDataSetChanged(State.IDLE);
                smartRefreshHelper.finishRefresh(true);
            }
        }, 2000);
    }

    @Override
    public void onLoadMore() {
        super.onLoadMore();
        smartRefreshHelper.getRecyclerView().postDelayed(new Runnable() {
            @Override
            public void run() {
                int size = modules.size();
                for (int i = size; i < size + 10; i++) {
                    modules.add("" + i);
                }
                if (modules.size() < 50) {
                    smartRefreshHelper.setLoadMore(true, false);
                } else {
                    smartRefreshHelper.setLoadMore(true, true);
                }
                adapter.notifyDataSetChanged();
            }
        }, 3000);
    }

    @Override
    protected BaseViewHolder onCreateBaseViewHolder(ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(SwipeRefreshLayoutPlusActivity.this).inflate(android.R.layout.simple_list_item_1, parent, false));
    }

    class Holder extends BaseViewHolder {
        private final TextView label;

        public Holder(@NonNull View itemView) {
            super(itemView);
            label = (TextView) itemView;
        }

        public void initData(int position) {
            String text = modules.get(position);
            label.setText(text);
        }
    }
}

File: app/src/main/java/com/qw/recyclerview/sample/ui/SwipeRefreshLayoutPlusActivity.kt
package com.qw.recyclerview.sample.ui

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import com.qw.recyclerview.core.adapter.BaseViewHolder
import com.qw.recyclerview.sample.R
import com.qw.recyclerview.sample.core.BaseSwipeRefreshLayoutListActivity
import com.qw.recyclerview.sample.loading.State

class SwipeRefreshLayoutPlusActivity : BaseSwipeRefreshLayoutListActivity<String>() {
    override fun setContentView() {
        setContentView(R.layout.swipe_refresh_layout_activity)
    }

    override fun initData(savedInstanceState: Bundle?) {
        loading.setOnRetryListener {
            for (i in 0..19) {
                modules.add("" + i)
            }
            adapter.notifyDataSetChanged()
            loading.notifyDataChanged(State.done)
        }
        loading.notifyDataChanged(State.ing)
        Handler().postDelayed({
            loading.notifyDataChanged(State.error)
        }, 2000)
    }

    override fun onRefresh() {
        smartRefresh.getRecyclerView().postDelayed({
            modules.clear()
            for (i in 0..19) {
                modules.add("" + i)
            }
            adapter.notifyDataSetChanged()
            smartRefresh.finishRefresh(false)
        }, 2000)
    }

    override fun onLoadMore() {
        smartRefresh.getRecyclerView().postDelayed({
            val size = modules.size
            for (i in size until size + 10) {
                modules.add("" + i)
            }
            if (modules.size < 50) {
                smartRefresh.finishLoadMore(success = true, noMoreData = false)
            } else {
                smartRefresh.finishLoadMore(success = true, noMoreData = true)
            }
            adapter.notifyDataSetChanged()
        }, 3000)
    }

    override fun onCreateBaseViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return object : BaseViewHolder(LayoutInflater.from(this@SwipeRefreshLayoutPlusActivity).inflate(android.R.layout.simple_list_item_1, parent, false)) {
            override fun initData(position: Int) {
                val label = itemView as TextView
                label.text = modules[position]
            }
        }
    }
}